server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -287,13 +298,13 @@ private HitContext prepareNonNestedHitContext(SearchContext context,
                 hit = new SearchHit(docId, fieldsVisitor.id(), emptyMap(), emptyMap());
             }
 
-            HitContext hitContext = new HitContext(hit, subReaderContext, subDocId);
+            HitContext hitContext = new HitContext(hit, valuesLookup, subReaderContext, subDocId);
             if (fieldsVisitor.source() != null) {
                 // Store the loaded source on the hit context so that fetch subphases can access it.
                 // Also make it available to scripts by storing it on the shared SearchLookup instance.
-                hitContext.sourceLookup().setSource(fieldsVisitor.source());
+                hitContext.valuesLookup().source().setSource(fieldsVisitor.source());
```

Comment:
This is still pretty ugly - would be nice to more stored fields access into `ValuesLookup` as well so that the lookup can handle access to everything and we don't need to set the source separately - but that's a much bigger change and is not necessary for the current project.